File: LeetCode 78 90-subset/duplicate_nums/subset_backtrack.py
class Solution(object):
    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = []
        nums.sort()
        
        def backtrack(idx, subset):
            if idx == len(nums):
                result.append(subset[::])
                return
            
            # subsets include nums[idx]
            subset.append(nums[idx])
            backtrack(idx+1, subset)
            subset.pop()
            
            # subsets exclude nums[idx]
            while idx + 1 < len(nums) and nums[idx] == nums[idx+1]:
                idx += 1
            backtrack(idx+1, subset)
        
        backtrack(0, [])
        return result
        
if __name__ == "__main__":
    solution = Solution()
    nums = [2, 1, 2]
    print("subsets with duplicate", solution.subsetsWithDup(nums))


# output: ('subsets with duplicate', [[1, 2, 2], [1, 2], [1], [2, 2], [2], []])

File: LeetCode 78 90-subset/duplicate_nums/subset_backtrack.go
package main

import (
	"fmt"
	"sort"
)

func subsetsWithDup(nums []int) [][]int {
	sort.Ints(nums)
	result := [][]int{}
	backtrack(nums, 0, []int{}, &result)
	return result
}

func backtrack(nums []int, idx int, subset []int, result *[][]int) {
	if idx == len(nums) {
		newSubset := append([]int{}, subset...)
		*result = append(*result, newSubset)
		return
	}

	// subsets include nums[idx]
	subset = append(subset, nums[idx])
	backtrack(nums, idx+1, subset, result)
	subset = subset[0 : len(subset)-1]

	// subsets exclude nums[idx]
	for idx+1 < len(nums) && nums[idx] == nums[idx+1] {
		idx += 1
	}
	backtrack(nums, idx+1, subset, result)
}

func main() {
	nums := []int{2, 1, 2}
	fmt.Println("subsets with duplicate", subsetsWithDup(nums))
}

// output: subsets with duplicate [[1 2 2] [1 2] [1] [2 2] [2] []]
